viewer/mlab-view

```diff
@@ -0,0 +1,176 @@
+  that the web server is shutdown when the operator's ssh connection has
+  disconnected.
+
+Example usage (from operator's workstation):
+
+  $ ssh -L 8088:utility.mlab.<hostname>:8088 \\
+      -p806 mlab_utility@<hostname> sudo mlab-view
+
+  Operator directs their browser to:
+
+      http://localhost:8088/
+
+What's happening?
+
+ * SSH binds to port 8088 on localhost.
+ * SSH login to <hostname> as the mlab_utility user and runs the command:
+
+     "sudo mlab-view".
+
+ * mlab-view starts a web server that listens on port 8088.
+ * The web server is configured to reject connections from any address other
+   than the slice IP.
+ * When ssh detects a connection to localhost:8088, the remote end attempts a
+   connection to:
+
+      "utility.mlab.<hostname>:8088"
+
+   To the server, this connection appears to come from the slice IP.
+ * SSH continues to forward all traffic back and forth from localhost:8088
+   through the ssh connection to the remote web server.
+"""
+
+import os
+import signal
+import subprocess
+import sys
+import syslog
+import time
+
+# Third-party modules.
+import netifaces
+
+
+ACCESS_CONF = '/usr/share/collectd-mlab/access.conf'
+HTTPD_PID = '/etc/httpd/run/httpd.pid'
+
+def log(msg):
+  """Sends log message to syslog."""
+  print msg
+  syslog.syslog(syslog.LOG_ERR, 'mlab-view: %s' % msg)
+
+
+def ignore_signal_handler(signal, frame):
+  """Handler to ignore any signal."""
+  log('Ignoring signal: %s' % signal)
+
+
+def get_address():
+  """Returns local IPv4 & IPv6 addresses as a string separated by spaces."""
+  ifaces = netifaces.interfaces()
+  addrs = []
+  for iface in ifaces:
+    if iface == 'lo':
+      continue
+    addresses = netifaces.ifaddresses(iface)
+    for addr_type in (netifaces.AF_INET, netifaces.AF_INET6):
+      if addr_type in addresses:
+        addrs.extend([addr['addr'] for addr in addresses[addr_type]])
+  if not addrs:
+    return '127.0.0.1'
+  return ' '.join(addrs)
+
+
+def get_access_config():
+  return """
+<Directory /usr/share/collectd/>
+    AddHandler cgi-script .cgi
+    DirectoryIndex bin/index.cgi
+    Order Deny,Allow
+    Allow from {ips}
+    Deny from all
+</Directory>
+""".format(ips=get_address())
+
+
+def find_last_parent(pid):
+  """Walks up process tree starting at pid, stopping at sshd or ppid == 1."""
+  try:
+    status_lines = open('/proc/%s/status' % pid, 'r').readlines()
+  except IOError:
+    return -1
+
+  # If the process name contains 'sshd', we'll stop there.
+  name_line = filter(lambda x: x.startswith('Name:'), status_lines)
+  if not name_line:
+    return -1
+
+  ppid_line = filter(lambda x: x.startswith('PPid:'), status_lines)
+  if not ppid_line:
+    return -1
+
+  ppid = int(ppid_line[0].split()[1])
+  if 'sshd' in name_line[0] or ppid == 1:
+    return ppid
+
+  # Continue up the tree.
+  return find_last_parent(ppid)
+
+
+def main():
+  signal.signal(signal.SIGINT, ignore_signal_handler)
```

Comment:
No problem. I'm not entirely clear how signals get routed from a client terminal to the remote process. These ignored signals are attempting to give mlab-view time to shutdown httpd.

The normal mode for mlab-view is when ssh does not allocate a pseudo-tty. In this mode, when a user presses Ctrl-c, the signal is delivered to the ssh command itself, and the corresponding sshd child process on the remote side simply exits. This lets mlab-view clean things up without any signals.

On the other hand, if ssh allocates a pseudo-tty (e.g. by passing "-t") and a user presses Ctrl-c, the signal is sent over the connection. Without ignoring the signal, mlab-view exits and httpd will happily keep running, which we don't want.

What's unclear to me is that ctrl-c (SIGINT) on the client eventually results in mlab-view receiving SIGTERM. So, this signal is ignored for that situation. But, since, I could not find a way to deliver SIGINT and mlab-view worked as expected without that signal ignored in the scenarios I checked, I removed it in: 0478c47103097060d0d40357b89c3aeae3f7bade

Is there a way to improve this?